src/snowflake/cli/api/project/schemas/entities/common.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import ABC
+from enum import Enum
+from typing import Generic, List, Optional, TypeVar
+
+from pydantic import AliasChoices, Field, GetCoreSchemaHandler, ValidationInfo
+from pydantic_core import core_schema
+from snowflake.cli.api.project.schemas.native_app.application import (
+    ApplicationPostDeployHook,
+)
+from snowflake.cli.api.project.schemas.updatable_model import (
+    IdentifierField,
+    UpdatableModel,
+)
+
+
+class EntityType(Enum):
+    APPLICATION = "application"
+    APPLICATION_PACKAGE = "application package"
```

Comment:
We have those strings as literal on entities classes and we have them here. Should we consider using this Enum as source of truth? Or should we use entities to build Enum / something?

Taking into account that we are constructing `Union[ApplicationEntity, ApplicationPackageEntity]` and
```py
_v2_entity_types_map = {
    EntityType.APPLICATION.value: ApplicationEntity,
    EntityType.APPLICATION_PACKAGE.value: ApplicationPackageEntity,
}
```

It means that adding new entity requires it to be registered in at least two places.


it may be wise to have

```py
KNOWN_ENTITIES = [
  ApplicationEntity,
  ApplicationPackageEntity
]

ENTITIES_TYPE = Union[*KNOWN_ENTITIES]
_v2_entity_types_map = {get_args(eval(inspect.get_annotations(e)["type"]))[0]: e for e in KNOWN_ENTITIES}
```

The dict comprehension is a bit of a magic, but probably there's something smarter we can do.

Reply:
Yes, that's what I tried at first but I couldn't make pydantic understand enums as discriminators and reverted to literal strings.

I'm not sure what I think about this dict comprehension, but I get the idea. I'll try to come up with a cleaner way to have a single source of truth.